src/main/java/org/broadinstitute/hellbender/engine/ReadsHtsgetDataSource.java

```diff
@@ -0,0 +1,391 @@
+     *         as configured via {@link #setTraversalBounds}, otherwise false
+     */
+    @Override
+    public boolean traversalIsBounded() {
+        return this.intervals != null || this.traverseUnmapped;
+    }
+
+    /**
+     * This data source can be queried even without index files
+     * @return always true
+     */
+    @Override
+    public boolean isQueryableByInterval() {
+        return true;
+    }
+
+    /**
+     * Iterate over all reads in this data source. If intervals were provided via {@link #setTraversalBounds},
+     * iteration is limited to reads that overlap that set of intervals.
+     *
+     * @return An iterator over the reads in this data source, limited to reads that overlap the intervals supplied
+     *         via {@link #setTraversalBounds} (if intervals were provided)
+     */
+    @Override
+    @Nonnull
+    public Iterator<GATKRead> iterator() {
+        return this.prepareIteratorsForTraversal(this.intervals, this.traverseUnmapped);
+    }
+
+    /**
+     * Query reads over a specific interval. This operation is not affected by prior calls to
+     * {@link #setTraversalBounds}
+     *
+     * @param interval The interval over which to query
+     * @return Iterator over reads overlapping the query interval
+     */
+    @Override
+    public Iterator<GATKRead> query(final SimpleInterval interval) {
+        return this.prepareIteratorsForTraversal(Collections.singletonList(interval), false);
+    }
+
+    /**
+     * @return An iterator over just the unmapped reads with no assigned position. This operation is not affected
+     *         by prior calls to {@link #setTraversalBounds}. The underlying file must be indexed.
+     */
+    @Override
+    public Iterator<GATKRead> queryUnmapped() {
+        return this.prepareIteratorsForTraversal(null, true);
+    }
+
+    private Iterator<GATKRead> prepareIteratorsForTraversal(final List<SimpleInterval> intervals, final boolean traverseUnmapped) {
+        final Map<SamReader, CloseableIterator<SAMRecord>> mapping = intervals == null
+            ? this.getReaders(traverseUnmapped)
+            : this.getReadersWithIntervals(intervals, traverseUnmapped);
+
+        if (mapping.isEmpty()) {
+            return Collections.emptyIterator();
+        }
+        final CloseableIterator<SAMRecord> samRecordIterator;
+        if (mapping.size() == 1) {
+            samRecordIterator = mapping.values().iterator().next();
+        } else {
+            final Set<SAMFileHeader> headers = mapping.keySet().stream().map(SamReader::getFileHeader).collect(Collectors.toSet());
+            samRecordIterator = new MergingSamRecordIterator(this.createHeaderMerger(headers), mapping, true);
+        }
+        this.iterators.add(samRecordIterator);
+        return new SAMRecordToReadIterator(samRecordIterator);
+    }
+
+    private Map<SamReader, CloseableIterator<SAMRecord>> getReaders(final boolean traverseUnmapped) {
```

Comment:
traverseUnmapped is confusingly named here.  The way I read it initially was that travereUnmapped mean "also include unmapped reads". I.e. traverseUnmapped == false implies NO unmapped unplaced reads and true means get all the reads including unplaced unmapped.  

The flag actually seems to behave as "onlyIncludeUnplacedUnmappedReads". I think it should be renamed and a comment would be useful.

Reply:
Yeah this behaviour is confusing, but I'm trying to follow what `ReadsPathDataSource` does which is more complicated.

If there are intervals and `traverseUnmapped == true` it returns the intervals and the unplaced unmapped reads
If there are no intervals and `traverseUnmapped == true` it returns only the unplaced unmapped reads.